### PR TITLE
control-service: add support for jfrog docker repo

### PR DESCRIPTION
**Pull request by dakodakov**
Add support for storing data job images in a 
jfrog repository.

### CODE REVIEW

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JfrogRegistryInterface.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+import org.jfrog.artifactory.client.Artifactory;
+import org.jfrog.artifactory.client.ArtifactoryClientBuilder;
+import org.jfrog.artifactory.client.model.AqlItem;
+import org.jfrog.filespecs.FileSpec;
+
+import java.util.List;
+
+/**
+ * This class is used to provide interface methods between an Amazon Elastic Container Registry and
```
Comment (mivanov1988) [inline projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JfrogRegistryInterface.java:21]: Amazon?

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java [hunk 2]
```diff
@@ -155,14 +153,42 @@ public boolean buildImage(
...
-        && dockerRegistryService.dataJobImageExists(imageName, credentials)) {
-      log.trace("Data Job image {} already exists and nothing else to do.", imageName);
-      return true;
+    if (registryType.equalsIgnoreCase(REGISTRY_TYPE_ECR)) {
+      var credentials = awsCredentialsService.createTemporaryCredentials();
+      builderAwsSecretAccessKey = credentials.awsSecretAccessKey();
+      builderAwsAccessKeyId = credentials.awsAccessKeyId();
+      builderAwsSessionToken = credentials.awsSessionToken();
+      awsRegion = credentials.region();
+      if ((actualDataJobDeployment == null
```
Comment (mivanov1988) [inline projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java:168]: This check appears in both cases, right? Can we change the if statements order?

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java [hunk 3]
```diff
@@ -155,14 +153,42 @@ public boolean buildImage(
...
+              || desiredDataJobDeployment
+                  .getPythonVersion()
+                  .equals(actualDataJobDeployment.getPythonVersion()))
+          && ecrRegistryInterface.checkEcrImageExists(imageName, credentials)) {
+        log.trace("Data Job image {} already exists and nothing else to do.", imageName);
+        return true;
+      }
+    } else {
+      builderAwsSecretAccessKey = REGISTRY_TYPE_GENERIC;
```
Comment (mivanov1988) [inline projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java:177]: Why is the generic not empty?

File: projects/control-service/projects/helm_charts/pipelines-control-service/values.yaml [hunk 4]
```diff
@@ -271,6 +271,13 @@ deploymentDockerRegistryUsernameReadOnly: ""
 deploymentDockerRegistryPasswordReadOnly: ""
 
 
+## [Required if deploymentDockerRegistryType=jfrog]
+## Connection settings for the Jfrog artifactory instance, used to check if a data job image already exists
+jfrogArtifactoryUrl: ""
```
Comment (mivanov1988) [inline projects/control-service/projects/helm_charts/pipelines-control-service/values.yaml:276]: It is not clear if these properties will be used exclusively for deployment. Alternatively, do we need separate properties for each registry type, or can we reuse the one for generic?

**Review by mivanov1988 — APPROVED** [review]

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DockerRegistryService.java [hunk 5]
```diff
@@ -45,9 +44,7 @@ public String builderImage() {
   }
 
   // TODO: Implement
-  public boolean dataJobImageExists(
-      String imageName, AWSCredentialsService.AWSCredentialsDTO awsCredentialsDTO) {
-    return registryType.equalsIgnoreCase("ecr")
-        && ecrRegistryInterface.checkEcrImageExists(imageName, awsCredentialsDTO);
+  public boolean dataJobImageExists(String imageName) {
+    return false;
```
Comment (mivanov1988) [inline projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DockerRegistryService.java:48]: why this method return false?

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JfrogRegistryInterface.java (hunk 1 above)
Comment (dakodakov) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JfrogRegistryInterface.java]: Fixed.

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java (hunk 2 above)
Comment (dakodakov) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java:168]: The code blocks where these checks are, also initialize variables and create ECR credentials, so I'd rather keep them as they are.

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java (hunk 3 above)
Comment (dakodakov) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java:177]: I would rather have a specific value in these variables, which are later on passed to the builders.

File: projects/control-service/projects/helm_charts/pipelines-control-service/values.yaml (hunk 4 above)
Comment (dakodakov) [reply projects/control-service/projects/helm_charts/pipelines-control-service/values.yaml:278]: Comment extended for clarification.

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DockerRegistryService.java (hunk 5 above)
Comment (dakodakov) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DockerRegistryService.java:48]: We extend the existing behavior to support Jfrog registry and ECR, but a generic Docker repository is still not supported.

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java [hunk 6]
```diff
@@ -155,14 +153,42 @@ public boolean buildImage(
...
+      builderAwsAccessKeyId = REGISTRY_TYPE_GENERIC;
+      builderAwsSessionToken = REGISTRY_TYPE_GENERIC;
+      awsRegion = REGISTRY_TYPE_GENERIC;
```
Comment (anitasingh1) [inline projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java:180]: all these aws variables we are not using in else block, do we need them here?

**Review by anitasingh1 — APPROVED** [review]

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java (hunk 6 above)
Comment (dakodakov) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/JobImageBuilder.java:180]: Yes, these need to be initialized as they are passed as arguments to the builder container.

**Pull request merged**